[AG-4640] Feature(tool-panel): allow tool panels built in and custom to render outside of the grid as a component in an external container

sdwvit opened the pull request:
Add external tool panel support and configuration options

 - Introduce `parent` property in SideBarDef to allow tool panels to render outside grid
 - Modify `openToolPanel()` api to accept second, optional `parent` parameter
 - Update docs
 - Disable cache for style synchronization in build configurations

Fixes [AG-4640](https://ag-grid.atlassian.net/browse/AG-4640)

Regular theme:
<img width="692" height="652" alt="Screenshot 2025-08-14 at 15 45 02" src="https://github.com/user-attachments/assets/b2cc69c0-3382-436f-91d9-116ab74eb022" />
As a modal window:
<img width="699" height="658" alt="Screenshot 2025-08-14 at 15 45 11" src="https://github.com/user-attachments/assets/bff2e153-fdf1-4ceb-9482-46dfadbfe39c" />
As a popup:
<img width="542" height="640" alt="Screenshot 2025-08-14 at 15 45 15" src="https://github.com/user-attachments/assets/73d23e20-a169-49af-9384-b166ad9e0235" />
As a sidedrawer:
<img width="427" height="708" alt="Screenshot 2025-08-14 at 15 45 18" src="https://github.com/user-attachments/assets/b0aab4c6-56c5-4364-a433-2ce12319ff6d" />
Legacy theme:
<img width="704" height="626" alt="Screenshot 2025-08-14 at 15 44 32" src="https://github.com/user-attachments/assets/f89ad9e6-4940-4b51-9ea4-5d16bd81f34a" />
Docs change:
<img width="870" height="793" alt="Screenshot 2025-08-19 at 11 52 50" src="https://github.com/user-attachments/assets/5d92347d-5db5-4204-b4f5-18e4910bf8cf" />
<img width="908" height="815" alt="Screenshot 2025-08-19 at 11 51 57" src="https://github.com/user-attachments/assets/acc6bc1a-d1b3-48fc-ae1a-05101d8fe9ef" />

sdwvit committed:
wip

sdwvit committed:
Remove CtrlsService dependency and add parent support for tool panels

- Removed CtrlsService dependency and wireBeans method
- Updated tool panel append logic to support external parent element
- Added 'parent' property to SideBarDef interface for custom positioning

sdwvit committed:
- Updated tool panel append logic to check for HTMLElement instance
- Applied theme classes to external parent elements via environment service

sdwvit committed:
- Introduce 'ag-external' class for external tool panel styling
 - Update theme selectors to include external panel styles
 - Disable cache for style synchronization in build configuration

sdwvit committed:
disable caching copy-styles

sdwvit committed:
add docs

sdwvit committed:
bump bundle size

peterjrreynolds commented on documentation/ag-grid-docs/src/content/docs/side-bar/index.mdoc:
```diff
@@ -86,6 +86,10 @@ The snippet above is demonstrated in the following example:
 {% /note %}
 
+{% note %}
+Set `popupParent` to `document.body` (or a dedicated container) to ensure all popups (e.g., column tool panel menus) are fully visible when using a tool bar parent option.
```
```suggestion
The [Popup Parent](./context-menu/#popup-parent) must be set to an element that contains both the tool parent parent and the grid to ensure all popups (e.g., Columns Tool Panel context menus) are fully visible.
```

peterjrreynolds commented on packages/ag-grid-enterprise/src/sideBar/agSideBar.ts:
```diff
@@ -332,7 +338,16 @@ class AgSideBar extends Component implements ISideBar {
-        this.appendChild(wrapperGui);
+
+        if (def.parent instanceof HTMLElement) {
+            this.environment.applyThemeClasses(def.parent);
```
Unless it's being used a lot, better for bundle size to just do `this.beans.environment` here and remove `wireBeans`

peterjrreynolds commented on packages/ag-grid-enterprise/src/sideBar/agSideBar.ts:
```diff
@@ -115,7 +121,7 @@ class AgSideBar extends Component implements ISideBar {
         if (openPanel.contains(activeElement)) {
             nextEl = _findNextFocusableElement(beans, openPanel, undefined, true);
-        } else if (isTargetUnderManagedComponent(openPanel, target) && backwards) {
+        } else if (isTargetUnderManagedComponent(openPanel, target)) {
```
`findFocusableElementBeforeTabGuard` is looking for items before the side bar, so we shouldn't be running this logic if tabbing forwards

sdwvit replied:
backwards is always true in this code, see [above](https://github.com/ag-grid/ag-grid/pull/11775/files#diff-de69bf8162e9c974756c0a530ece8286a4bbc9b03a8f3dd9216b56ab44173619R109)

peterjrreynolds commented on packages/ag-grid-enterprise/src/sideBar/agSideBar.ts:
```diff
@@ -332,7 +338,16 @@ class AgSideBar extends Component implements ISideBar {
+            this.environment.applyThemeClasses(def.parent);
+            wrapperGui.classList.add(this.gos.get('enableRtl') ? 'ag-rtl' : 'ag-ltr');
+            def.parent.classList.add('ag-external');
+            def.parent.classList.add('ag-tool-panel-external');
```
`def.parent.classList.add('ag-external', 'ag-tool-panel-external');`

peterjrreynolds commented on packages/ag-grid-enterprise/src/sideBar/agSideBar.ts:
```diff
@@ -332,7 +338,16 @@ class AgSideBar extends Component implements ISideBar {
+            wrapperGui.classList.add(this.gos.get('enableRtl') ? 'ag-rtl' : 'ag-ltr');
+            def.parent.classList.add('ag-external');
+            def.parent.classList.add('ag-tool-panel-external');
+            def.parent.appendChild(wrapperGui);
```
Bundle size - `def.parent` used multiple times

peterjrreynolds commented on documentation/ag-grid-docs/src/content/docs/side-bar/index.mdoc:
```diff
@@ -86,6 +86,10 @@ The snippet above is demonstrated in the following example:
 Calling `setSideBarVisible(true)` when `sideBarDef.hideButtons` is set to true and no tool panel is open will not display anything.
 {% /note %}
 
+{% note %}
```
Should also add a docs example here demonstrating the feature

sdwvit committed:
Update documentation/ag-grid-docs/src/content/docs/side-bar/index.mdoc

Co-authored-by: Pete Reynolds <40919976+peterjrreynolds@users.noreply.github.com>

Commit:
- Updated applyThemeClasses to accept additional CSS classes

Signed-off-by: Victor Musienko <victor.m@noibu.com>

Commit:
Add Tool Panel Parent Configuration and Example

- Added documentation for `toolPanelDef.parent` property to render tool panels outside the side bar
- Created new example with HTML, TS, and CSS files to demonstrate custom tool panel positioning
- Implemented code snippet showing `parent: document.getElementById('toolPanelParent')` usage
- Updated example to include modal/drawer functionality for tool panel rendering
- Added styles for modal/drawer layout and component visibility control

Signed-off-by: Victor Musienko <victor.m@noibu.com>

sdwvit committed:
- Updated documentation to clarify rendering Tool Panels outside the grid
 - Added popupParent property to gridOptions for external rendering

sdwvit committed:
update docs

sdwvit committed:
update example

sdwvit committed:
Add parent parameter to openToolPanel for custom positioning

 - Update openToolPanel method signature in GridApi and ISideBar interfaces to include optional parent element
 - Implement logic in agSideBar.ts to handle parent element when creating tool panels
 - Update documentation example to demonstrate attaching tool panel to custom container
 - Add parent parameter to sideBarApi.ts openToolPanel export for consistent API

sdwvit committed:
update label

peterjrreynolds commented on documentation/ag-grid-docs/src/content/docs/side-bar/index.mdoc:
```diff
@@ -192,6 +196,35 @@ If you are using the long form (providing a `SideBarDef` object) then it is poss
+
+{% gridExampleRunner title="Tool Panel Parent" name="tool-panel-parent" exampleHeight=600 /%}
+
+The example above demonstrates displaying the Columns Tool Panel in a modal. Popup Parent is set to the document body. Following configuration is used: 
```
This needs to be updated to explain both the API method and the parent property in the def

sdwvit replied:
<img width="933" height="1028" alt="Screenshot 2025-08-21 at 15 41 42" src="https://github.com/user-attachments/assets/e67a7bca-533d-45cc-9bd4-98c6476fc955" />
This looks fine?

peterjrreynolds commented on documentation/ag-grid-docs/src/content/docs/side-bar/_examples/tool-panel-parent/main.ts:
```diff
@@ -0,0 +1,86 @@
+        return;
+    }
+    const modalDrawer = document.createElement('div');
+    modalDrawer.innerHTML = `
```
We will need provided versions of this example for the other frameworks as React, etc. shouldn't be doing direct DOM manipulation

peterjrreynolds commented on packages/ag-grid-community/src/api/gridApi.ts:
```diff
@@ -1307,9 +1307,10 @@ export interface _SideBarGridApi<TData> {
      * @agModule `SideBarModule`
      */
-    openToolPanel(key: string): void;
+    openToolPanel(key: string, parent?: Element | null): void;
```
```suggestion
    openToolPanel(key: string, parent?: HTMLElement | null): void;
```

peterjrreynolds commented on packages/ag-grid-community/src/interfaces/iSideBar.ts:
```diff
@@ -12,7 +12,11 @@ export interface ISideBar {
+    openToolPanel(
+        key: string,
+        source?: 'sideBarButtonClicked' | 'sideBarInitializing' | 'api',
+        parent?: Element | null
```
```suggestion
        parent?: HTMLElement | null
```

peterjrreynolds commented on packages/ag-grid-enterprise/src/sideBar/agSideBar.ts:
```diff
@@ -347,8 +351,24 @@ class AgSideBar extends Component implements ISideBar {
         key: string | undefined,
-        source: 'sideBarButtonClicked' | 'sideBarInitializing' | 'api' = 'api'
+        source: 'sideBarButtonClicked' | 'sideBarInitializing' | 'api' = 'api',
+        parent?: Element | null
```
```suggestion
        parent?: HTMLElement | null
```

peterjrreynolds commented on packages/ag-grid-enterprise/src/sideBar/sideBarApi.ts:
```diff
@@ -13,8 +13,8 @@ export function setSideBarPosition(beans: BeanCollection, position: 'left' | 'ri
 
-export function openToolPanel(beans: BeanCollection, key: string) {
-    beans.sideBar?.comp.openToolPanel(key, 'api');
+export function openToolPanel(beans: BeanCollection, key: string, parent?: Element | null) {
```
```suggestion
export function openToolPanel(beans: BeanCollection, key: string, parent?: HTMLElement | null) {
```

sdwvit committed:
Update openToolPanel to use HTMLElement for parent parameter

- Changed type from Element to HTMLElement in gridApi.ts and iSideBar.ts
- Updated tool-panel-parent example to use HTMLElement for modal drawer
- Adjusted event handling in documentation examples to manage parent container
- Ensured consistency across side bar API implementations

sdwvit committed:
Adjust modal drawer styles and event handling

- Update modal drawer styles for better visibility with theme variables
- Refactor toggleDrawer and passModal functions for improved modal management
- Remove inline onclick handler from modal drawer container

sdwvit committed:
Refactor tool panel rendering and parsing to support custom parent elements

- Added `renderToolPanelUnderParent` to delegate tool panel rendering to specified parent element
- Modified `openToolPanel` to use delegated rendering via new method
- Introduced `parseOneComponent` in parser to handle string-based tool panel definitions
- Updated documentation to describe parent property usage and API method for custom rendering
- Simplified inline parsing logic in side bar component

peterjrreynolds commented on documentation/ag-grid-docs/src/content/docs/side-bar/index.mdoc:
```diff
@@ -192,6 +196,19 @@ If you are using the long form (providing a `SideBarDef` object) then it is poss
 
+### Tool Panel Parent
+
+By default, Tool Panels are rendered inside the Side Bar. If you want to render Tool Panels in a different location, you can set the [parent](https://localhost:4610/javascript-data-grid/side-bar#reference-ToolPanelDef-parent) property in the `ToolPanelDef`. This is useful if you want to render Tool Panel in a different part of your application, such as a popup window or a separate section of your ...
```
URL path

sdwvit committed:
- Correct link to the parent property in ToolPanelDef documentation
